### ISSUE TITLE
rl4j - frequent syntax errors

### ISSUE THREAD

**Issue by kgoderis** [deeplearning4j/deeplearning4j#9594]
I was wondering how the rl4j can be compiled because it is seemingly full of syntax errors. For example, 

https://github.com/eclipse/deeplearning4j/blob/aa1961623bccedc6ddcb3973b0b42af841417afb/rl4j/rl4j-api/src/main/java/org/deeplearning4j/rl4j/space/DiscreteSpace.java#L27 should implement https://github.com/eclipse/deeplearning4j/blob/aa1961623bccedc6ddcb3973b0b42af841417afb/rl4j/rl4j-api/src/main/java/org/deeplearning4j/rl4j/space/ActionSpace.java#L32 but it doesn't

Importing the mvn project in an IDE will reveal many similar issues. The list is quite endless (not taking into account that there is a lot of Lombok going on)

**Comment by kgoderis** [comment]
Lombok!

**Comment by kgoderis** [comment]
Eclipse is behaving weird

**Comment by treo** [comment]
If anyone finds this and is wondering: This is due to our use of https://projectlombok.org/ 

https://github.com/eclipse/deeplearning4j/blob/aa1961623bccedc6ddcb3973b0b42af841417afb/rl4j/rl4j-api/src/main/java/org/deeplearning4j/rl4j/space/DiscreteSpace.java#L30-L31

This implements `getSize()` with Lombok's `@Getter` annotation. 

There is a plugin for most IDE's that will help it understand all of that. e.g. eclipse: https://projectlombok.org/setup/eclipse

**Comment by kgoderis** [comment]
Indeed, but somehow in Eclipse not all errors are resolved even after installing the Lombok plugin

**Comment by kgoderis** [comment]
@treo The rl4j code base is difficult to understand because of all the depreciated classes that are also used by the examples. Just for my understanding, what was the reason to depreciate classes and what are the alternatives classes that need to be used? This is not clear, or is there are different strategy here?

For example, what is the difference between org.deeplearning4j.rl4j.agent.learning.algorithm.actorcritic.* and org.deeplearning4j.rl4j.learning.async.a3c.*

or, org.deeplearning4j.rl4j.mdp and org.deeplearning4j.rl4j.environment ? 

It looks like there are two implementations in rl4j, with all the example referring to the classes that contain all these depreciated methods

**Comment by agibsonccc** [comment]
@kgoderis 


Sorry there hasn't been a ton of time to maintain rl4j. I'm actually thinking of just deprecating it completely. I've left it there because some people seem to use it but internally we don't have really touch it much. 



I've already purged a number of modules for our upcoming 1.0 release. If you would like to step up and contribute I think we could guide you on at least helping with some of the clean up.

Given your frustration here I'm erring on doing just doing the deprecation. RL4j was mostly  a community maintained module we started a few years ago but haven't been able to put the proper effort in to. 

We were originally going to do something with it a few years ago but that interest died down a while back.

Usually when folks have issues like this if we aren't maintaining it internally and no one in the community is willing to step in (which is usually the case) then it's easier for us to just maintain what we actually use ourselves while the community benefits from whatever we find useful for our own team. At least then interests are aligned and it isn't a burden on us to maintain what's there.

**Comment by kgoderis** [comment]
@agibsonccc i understand fully and I am ok to contribute as I go along, and i am equally ok to do a cleanup. I was just wondering what parts are still relevant, what is dated, but from what I discovered yesterday I sense that the Maze example and the classes that are used by that example are the most robust and elaborated

Just wondering, did interest falter because of mediocre RL performance? RL's share of voice has been increasing year over year

**Comment by agibsonccc** [comment]
@kgoderis after chatting with @treo we agreed that a standalone tutorial for the samediff and deeplearning4j DL APIs would be easier. The main problem is RL is more of a research thing and not used much in industry. The simulation community would be one of the main adopters of this technology but they are behind a bit and do well enough with their standalone techniques so it's more of a commercial angle. 

We're not against RL in principal it was just the ROI of the rl4j framework itself. It was written a few years ago as a research project with the intention of going to production but it ended up requiring a lot of up front refactoring.

The good news is the necessary tools to make it work like external errors are built in to dl4j and samediff and I have seen users do fine with just that. If you want maybe we could collaborate on a new guide on how to apply external errors to a reinforcement learning problem.

**Comment by kgoderis** [comment]
@agibsonccc @treo.AFAIK, but I am still discovering, there is no other RL java framework with any good traction or that is backed with core functionalities like DL4J or SameDiff, so for me RL4J remains a bit the go-to option.  I understand that at the end industry is more interested in putting the agent networks into production, rather than try out new algo's. 

I am not sure I fully understand what you mean with external errors here, but if you refer to Integrating DL4J designed networks,  also feel this is quite doable and smooth. I will also assume that anything that is @Depreciated in the RL4J code base is really depreciated and can be safely removed at some point in time. 

According to you, what would be the attention points that need fixing in RL4J? (maybe you have an old roadmap floating around that can serve as the base for any additional efforts from the community)

**Comment by agibsonccc** [comment]
@kgoderis basically loss functions are derived from policies (think rewards) which are "external".

Roadmap wise if you want to step up:

1. More flexibility is needed. The current architecture gets in the way.
2. More algorithms are needed ideally. It's a bit behind the current state of the art. Model import work we are doing could fix this somewhat.

3. Better data pipelines and scenario building similar to open ai gym. We had an integration with that at one point.
4. Code base clean up: we were in the middle of a refactoring and never got back to it. @aboulang2002  was doing some changes to it for a while alongside @saudet 
but again time/interest was the main factor. You can see what was being worked on last here: https://github.com/eclipse/deeplearning4j/pulls?q=is%3Apr+is%3Aclosed+author%3Aaboulang2002

**Comment by treo** [comment]
Ideally we would take a good hard look at RL4J and the use-cases upfront, in particular the part you mention:
>  I understand that at the end industry is more interested in putting the agent networks into production

I find RL4J in its current state to be targeted too much at the other end of that spectrum while at the same time not actually having anything really worthwhile there.

For the most part, I think that you can easily get an agent network into production without any support by RL4J at all. If you have a trained policy that takes an observation as its input and will return you the next action - then that is literally just inference of a trained model.

For training, RL4J in its current state is quite hard to reasonably use. The state of it isn't significantly different than it was when I created the simple snake example (https://github.com/treo/rl4j-snake), so only the simplest things really work there. 

If we can work out use cases beyond game-playing and find an architecture for it that is both easily understandable and extensible then it may be worth the effort - but I think it will not look too much like RL4J currently does anymore. 

If I remember correctly, one of the reasons that slowed down the previous efforts of refactoring RL4J was the intent to keep existing code working.

**Comment by kgoderis** [comment]
> For the most part, I think that you can easily get an agent network into production without any support by RL4J at all. If you have a trained policy that takes an observation as its input and will return you the next action - then that is literally just inference of a trained model.

Indeed, but it is also about updating a network with new data, or hyperparameter tuning a la Arbiter.

> 
> For training, RL4J in its current state is quite hard to reasonably use. The state of it isn't significantly different than it was when I created the simple snake example (https://github.com/treo/rl4j-snake), so only the simplest things really work there.
> 

I looked at that example, and it using the methods and classes that are @depreciated, but I can imagine some for the initial refactoring happened between now and 14 months ago. 

> If we can work out use cases beyond game-playing and find an architecture for it that is both easily understandable and extensible then it may be worth the effort - but I think it will not look too much like RL4J currently does anymore.

To me the refactored stuff already made a lot of sense in terms of architecture

> 
> If I remember correctly, one of the reasons that slowed down the previous efforts of refactoring RL4J was the intent to keep existing code working.

Well, the worst thing one can do is keep the technical debt; I guess you are ok to make code breaking changes

**Comment by aboulang2002** [comment]
Hello @kgoderis, 
Yes, as you have suspected, there are two versions of RL working side by side. Originally, I didn't know who was using RL4J so I thought that it was important to keep the legacy code working, even at the cost of code clarity.

My original aim was to be able to do '5 minutes prototypes'; to make it possible for someone to come up with an RL idea and to be able to code it within 5 minutes. As I implemented features, something became very clear: RL algorithms may looks similar, but they have a lot of small differences. And to remain as generic as possible means that you are forced to divide the code in a lot of tiny classes/methods. In the end you are almost assembling single lines of code. As the number of interfaces and classes grow, the library become less and less friendly to use and also more and more difficult to keep a good mental image of the architecture. 

In hindsight, I'm not convinced that creating a RL library is worth it. If I had a recommendation to make, it would be to make reference reinforcement learning implementations with DL4J, similar to what OpenAI did.


If you still want to work on RL4J, there are code examples in the tests. I used these to demonstrate my changes worked as intended and did not cause any regression:
	See `deeplearning4j/rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/ `

- `AgentLearnerCartpole.java`: Uses the cartpole environment and learns with DoubleDQN, NStepQLearning, and A2C/A3C
- `NStepRnn.java`: Used to test RL4J with RNNs. See https://github.com/KonduitAI/deeplearning4j/pull/531
- `RobotLakeExample.java`: Demonstrate how an environment can return multiple data channels. See https://github.com/KonduitAI/deeplearning4j/pull/548
- `TMazeExample.java`: Also used to test RNNs. See https://github.com/KonduitAI/deeplearning4j/pull/514

The general idea is this:

- The environment generates a new state
- That state is fed through a transformation pipeline to convert it into an observation that can be used by RL4J (Search transformProcess in `rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/agent/Agent.java`)
- That observation is then fed through a policy to determine the next action
- The environment is stepped with that action to obtain the reward and the next observation
- These are sent to the learning behavior (see `onAfterAction` of `rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/agent/AgentLearner.java`)
- The learning behavior then gives the experience to the experience handler (how an experience is handled is dependant on the algorithm used (DDQN, A3C, ...))
- When enough experience has been accumulated to perform an update, the updateRule is called.

	
I hope this is enough to get you started.

**Comment by kgoderis** [comment]
@aboulang2002 Yes, thank you for this elaborate explanation. In the meantime I have started to dive deeper in the code. So far I found the architecture quite clean, in fact, and certainly not something that should be thrown away or abandoned. The only thing that I found a bit limiting so far is the fact that the classes are built around Action(s) that are Integer, and not a POJO. E.g. you could have Agents that return something else than an softmaxed result, or, Actions might need some preprocessing before they fed to an Environment (for example, combine it with the outcome of another agent in a collaborative setting)

**Comment by kgoderis** [comment]
> @kgoderis basically loss functions are derived from policies (think rewards) which are "external".
> 
> Roadmap wise if you want to step up:
> 
> 1. More flexibility is needed. The current architecture gets in the way.

@agibsonccc I have studied the library (e.g. the refactored bits) over the weekend, and I found it reasonably flexible. The usage of Configuration classes and Builders is quite extensive and does bring flexibility. In which regard does the architecture get in the way? Do you want to reduce the usage of templated classes ? ... just wondering 

K

**Comment by agibsonccc** [comment]
@kgoderis it mainly comes down to the architectures you can specify. 
If you look we typically just allow feed forward and rnns.

We allow a very limited number of nets compared to what you might expect. Ideally it would be more extensible.

I appreciate you stepping up with the work you're doing! I look forward to the result.

**Comment by kgoderis** [comment]
> We allow a very limited number of nets compared to what you might expect. Ideally it would be more extensible.

@agibsonccc Can you give an example of what kind of networks we should be able to support? This will help me understand how to generalise what is in rl4j, e.g. step away from various Helper class implementations to something more future proof.

**Comment by agibsonccc** [comment]
@kgoderis Generally just being able to specify an arbitrary MultiLayerNetwork or ComputationGraph, or even SameDiff instance (maybe with a dl4j wrapper) over the current api would be nice. Right now it just restricts you and creates the network internally.

**Comment by kgoderis** [comment]
> @kgoderis Generally just being able to specify an arbitrary MultiLayerNetwork or ComputationGraph, or even SameDiff instance (maybe with a dl4j wrapper) over the current api would be nice. Right now it just restricts you and creates the network internally.

Can one assume that the dimensionality/dimensions of the features and labels fed into the network are the constant in the sense that het network adjusts itself to the algorithm, and not the other way around, or do you reckon we need this to be full flexible?

**Comment by treo** [comment]
It depends on what kind of RL algorithm ends up being used. But in general the Observation and ActionSpace of an MDP provide the information for the input and output size. 

However, I think that it doesn't necessarily make sense to constrain the possible models because of this. A simple compatibility check for inputs and outputs of the provided model is probably going to be just as ergonomic. 

One thing that I'm not 100% sure about is whether RL4J needs to be able to produce new empty models for some RL algorithms and in that case providing a pre-existing model obviously doesn't work, but I'd still argue for accepting a producer.

**Comment by kgoderis** [comment]
@treo I misformulated my question a bit, but in fact, in org.deeplearning4j.rl4j.network there are the classes that can handle the format of the inputs/outputs of the networks. I am bit puzzled by the (@agibsonccc ) on the network flexibility, as it seems - to me - that everything is already in that package to support any type of neural net. There are Factory classes for AC and DQN based networks but they are deprecated/can be maybe salvaged

On a separate note, to what extend are you keen to keep all the Generics in the framework ? There are quite a bunch of them, and they are not all that helpful IMO. I would remove them, push some logic into classes and extend the Builders to be able to handle variability. For example, first thing for me would be to generalise the actions into an Action class (iso Integer everywhere), build an hierarchy of classes (DiscreteAction, ContinuousAction,...), that are used in ActionSpace | ActionSchema, serving as factory to Actions (to solve potential issue of using reflection in order to instantiate a Generic class).

**Comment by treo** [comment]
I think part of this is probably that both @agibsonccc and me may have mostly the API in mind that existed before @aboulang2002 started the refactoring. 

For me at least, I may have a somewhat outdated opinion in that regard.

> On a separate note, to what extend are you keen to keep all the Generics in the framework ? There are quite a bunch of them, and they are not all that helpful IMO

If it makes it harder to use or harder to understand, then I think then generics aren't really helping. I also remember there being a lot of generic usage for not a lot of benefit too, but again I may remember an older state of the library.

**Comment by agibsonccc** [comment]
Yeah I think @treo is right upon looking at it. If generics hinder the usage they don't need to be there.

**Comment by aboulang2002** [comment]
@kgoderis Not a lot have been done regarding building the networks, Please see https://github.com/KonduitAI/deeplearning4j/pull/514#pullrequestreview-477664485, where some ideas have been discussed.

**Comment by kgoderis** [comment]
@aboulang2002 Thx for that input. Building networks is indeed one thing, but it is correct to state you already have built a lot of functionality to inject networks into the RL algos, right? I don't think that part needs any changes, and it is a big improvement over the legacy code

**Comment by aboulang2002** [comment]
@kgoderis Correct; any DL4J network can be used in the new code of RL4J.

**Comment by agibsonccc** [comment]
@kgoderis since I'm not seeing much in terms of actual activity I'm moving rl4j out of the main repo. We are doing a java 9+ upgrade and overhauling some modules. It might not be a had idea to fork or potentially create a repo we can link to. If you're up for that I'd be happy to support that and link it on the site.

**Comment by kgoderis** [comment]
@agibsonccc FYI, I have done quite some work, in a forked repo, but it is not ready to be pushed/merged. (see https://github.com/kgoderis/deeplearning4j/tree/rl4j-refactor)

**Comment by kgoderis** [comment]
@agibsonccc What I have done so far
 - Introduce generalist Action and Observation classes
 - Migrate all legacy examples into the main repo (so, ALE, Doom,... and rl4j-examples) + refactor
 - Clean-up the regular examples
 - Get rid of legacy code
 
Still to do : update Unit tests, further clean-up

**Comment by agibsonccc** [comment]
@kgoderis thanks a lot. I'm not against supporting this in some capacity, let's just figure out how to move this out of the main repo. I'd be happy to help supporting in publishing it. I just want to make sure we keep the dl4j library itself small and easier to manage.

One thing we can definitely do is keep the examples active there.

**Comment by kgoderis** [comment]
I did the reverse and moved all examples into the rl4j-core project so that everything sits together, making it a bit easier for people to find everything at one place, and make sure the examples will not get out of sync with the core rl4j classes. 

I am by no means a developer by profession, so some parts might need some help but I will indicate that (at least adding the proper unit tests will be an issue). It will need review; I tried to build upon what @aboulang2002 did in the past.

**Comment by kgoderis** [comment]
@agibsonccc @aboulang2002 @treo All, I have just created https://github.com/kgoderis/deeplearning4j/tree/rl4j-refactor in which I have now consolidated a bunch of changes to RL4J. These include the removal of all legacy code, further generalisation of previous work, a refactoring of all examples to use the refactored code and so forth. In the end I have decided to keep the generics approach; I also combined all modules into a single repo just for dev purposes, this might need to be split out again in the future. The examples do run, the only I could not full test is ALE as the ALE interpreter does not work well with the pong.bin ROM I found on the net. Apart from that, it still needs further code beautification, and more importantly, all the Unit Tests need to be adjusted. I will need help with that last bit - I am not a dev by profession, and never really did test-driven development in my life. (e.g. learning curve to climb here, and don't have a lot of bandwidth to do that). I did not yet implement a new algo (PPO?) as a proof of case

I would be very happy if you can have a glance at the code and share some thoughts. I noticed that in the meantime rl4j was removed from the main repo, so I am happy to discuss how rl4j can be published and all of that.

**Comment by agibsonccc** [comment]
@kgoderis yes happy to help. Maybe migrate this over to our forums for some visibility: https://community.konduit.ai/ - M2 has been published and now we are able to help with this a bit.

For release support we are thinking of using something like our contrib folder. For publishing, we can try to put it under the org.eclipse.deeplearning4j namespace with maybe an extra .contrib on the end of that for 3rd party modules. Again though let's discuss that more on the forums. I'll  be fine with supporting whatever you need so people benefit from the work you did.

**Comment by kgoderis** [comment]
> For release support we are thinking of using something like our contrib folder. For publishing, we can try to put it under the org.eclipse.deeplearning4j namespace with maybe an extra .contrib on the end of that for 3rd party modules. Again though let's discuss that more on the forums. I'll be fine with supporting whatever you need so people benefit from the work you did.

That's perfect for me. Will move and open the debate ;-)

**Comment by kgoderis** [comment]
@agibsonccc Adam, shall I press on and do a mass renaming to org.eclipse.deeplearning4j.contrib.rl4j, move to contrib, and prepare a PR? Shall I keep al test code, or simply remove that? It is not update, and would require heavy changes and additions, something that I am not familiar with? Keeping it could mean things don't run well in the CI/CD pipelines

**Comment by agibsonccc** [comment]
@kgoderis not quite...I still need to sort out how the next release will work and how the contrib folder will come in to play. Feel free to ping me on the forums as well I'll see that. Sorry just been busy with other things.
